feat(Scroll): Decoder for transactions bridging to and from Scroll

zinkkrysty opened the pull request:
Includes [refactor(OptimismSuperchain): rename to L2WithL1Fees](https://github.com/rotki/rotki/commit/7b713302a4cf71fa6a1da02d854cd2ddeea801c3)

zinkkrysty commented on rotkehlchen/chain/evm/decoding/utils.py:
```diff
@@ -50,7 +50,9 @@ def maybe_reshuffle_events(
 def bridge_prepare_data(
         tx_log: 'EvmTxReceiptLog',
         deposit_topics: Sequence[bytes],
+        # The chain of the current transaction
         source_chain: ChainID,
+        # The chain on the other side
         target_chain: ChainID,
```
it took me a long time to get what `source` and `target` meant here. I was understanding them as `from` and `to` and didn't make sense when decoding on the side of the scroll chain. So I added some comments to clarify

yabirgb commented on rotkehlchen/chain/ethereum/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,369 @@
+
+        # Find the fee refund event and remove it
+        refund_event_idx, refund_amount = None, FVal(0)
+        for idx, event in enumerate(context.decoded_events):
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.location_label == from_address and
+                event.asset == self.eth and
+                event.address == L1_MESSENGER_PROXY
+            ):
+                refund_event_idx = idx
+                refund_amount = event.balance.amount
+                break
+        if refund_event_idx is not None:
+            del context.decoded_events[refund_event_idx]
+
+        # Find the corresponding transfer event and update it
+        for event in context.decoded_events:
```
you are iterating twice the events. Can'y you di it in one loop and keeping the id of the refund event to remove it later?

yabirgb commented on rotkehlchen/chain/ethereum/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,369 @@
+                refund_event_idx = idx
+                refund_amount = event.balance.amount
+                break
+        if refund_event_idx is not None:
+            del context.decoded_events[refund_event_idx]
```
why do you need to remove an event?

yabirgb commented on rotkehlchen/chain/ethereum/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,369 @@
+        )
+
+        # Find the fee refund event and remove it
+        refund_event_idx, refund_amount = None, FVal(0)
```
```suggestion
        refund_event_idx, refund_amount = None, ZERO
```

yabirgb commented on rotkehlchen/chain/ethereum/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,369 @@
+                l2_fees_amount = event.balance.amount - refund_amount
+                event.counterparty = CPT_SCROLL
+                event.balance.amount = l2_fees_amount
+                event.notes = (
+                    f'Spend {l2_fees_amount} ETH in L2 fees to bridge ERC20 tokens to Scroll'
+                )
```
in general in those cases we prefer to use # noqa: E501 rathern than creating the string with this syntax

yabirgb commented on rotkehlchen/chain/ethereum/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,369 @@
+            to_chain: ChainID,
+    ) -> EvmEvent:
+        """Adjusts the original decoded event to an ETH bridge event"""
+
```
we try to not leave empty lines like this one

```suggestion
```

yabirgb commented on rotkehlchen/chain/l2_with_l1_fees/types.py:
Maybe it would be better to place this function in the evm/types.py file

zinkkrysty replied:
Not sure I understand this. Could you elaborate why this would be better? These are specific to only `L2WithL1Fees` type of chains, shouldn't these be in a place separate from the main EVM functionality?

yabirgb reviewed (commented):
Made a few comments. There are tests failing that would need attention

zinkkrysty commented on rotkehlchen/chain/scroll/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,232 @@
+
+L2_ETH_GATEWAY: Final = string_to_evm_address('0x6EA73e05AdC79974B931123675ea8F78FfdacDF0')
+L2_ERC20_GATEWAY: Final = string_to_evm_address('0xE2b4795039517653c5Ae8C2A9BFdd783b48f447A')
+# USDC has a special gateway
+L2_USDC_GATEWAY: Final = string_to_evm_address('0x33B60d5Dd260d453cAC3782b0bDC01ce84672142')
```
I got through randomly browsing transactions on scrollscan
But I can see now it's in their official repo as well: 

https://github.com/scroll-tech/scroll/blob/ed0e0e4c188cca1fd2a3ae71a0be03220be0050f/bridge-history-api/conf/config.json#L34

zinkkrysty commented on rotkehlchen/chain/scroll/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,232 @@
+            ['address', 'address', 'uint256'],
+            method_input_data,
+        )
+        from_address = string_to_evm_address(to_checksum_address(from_hex))
+        to_address = string_to_evm_address(to_checksum_address(to_hex))
```
no unfortunately, they are all lowercase addresses. I'm not sure if `decode_abi` has a method for decoding checksummed addresses from bytes. `hex_or_bytes_to_address` also calls `to_checksum_address`  and we call that quite often in other bridge decoding methods.

zinkkrysty commented on rotkehlchen/chain/ethereum/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,369 @@
+            sender=user_address,
+        )
+
+        if refund_event_idx is None:
+            return DEFAULT_DECODING_OUTPUT
```
I don't know if this can happen in a real scenario... I will update it so that I handle that scenario. Basically we need to replace the refund event with the fees event that we create, so in case the refund event won't be found (refund is 0), we can just add the fee event.

zinkkrysty commented on rotkehlchen/globaldb/updates.py:
```diff
@@ -35,9 +35,10 @@
-ASSETS_UPDATES_URL = 'https://raw.githubusercontent.com/rotki/assets/{branch}/updates/{version}/updates.sql'
-ASSET_COLLECTIONS_UPDATES_URL = 'https://raw.githubusercontent.com/rotki/assets/{branch}/updates/{version}/asset_collections_updates.sql'
-ASSET_COLLECTIONS_MAPPINGS_UPDATES_URL = 'https://raw.githubusercontent.com/rotki/assets/{branch}/updates/{version}/asset_collections_mappings_updates.sql'
+# https://raw.githubusercontent.com/zinkkrysty/rotki-assets/scroll-usdc/updates/24/updates.sql
+ASSETS_UPDATES_URL = 'https://raw.githubusercontent.com/zinkkrysty/rotki-assets/scroll-usdc/updates/{version}/updates.sql'
+ASSET_COLLECTIONS_UPDATES_URL = 'https://raw.githubusercontent.com/zinkkrysty/rotki-assets/scroll-usdc/updates/{version}/asset_collections_updates.sql'
+ASSET_COLLECTIONS_MAPPINGS_UPDATES_URL = 'https://raw.githubusercontent.com/zinkkrysty/rotki-assets/scroll-usdc/updates/{version}/asset_collections_mappings_updates.sql'
```
yeah, sorry for this. mistakenly committed them.

zinkkrysty commented:
@yabirgb I have addressed most of the comments: see commit description for https://github.com/rotki/rotki/pull/7698/commits/b009b3bbc4543da1d3a252eb9db5c7ad46b90852

Next I will address the failing tests, but I need to take a break now.

zinkkrysty commented:
I didn't get the time to address the failing specs. I will look into them hopefully later today. It seems a bunch of decoder tests fail because they use `EvmTransaction` instead of the new one with the `l1_fee`.

LefterisJP commented on rotkehlchen/chain/l2_with_l1_fees/types.py:
```diff
@@ -0,0 +1,71 @@
+    Timestamp,
+)
+
+L2_CHAINS_WITH_L1_FEES = [
+    SupportedBlockchain.OPTIMISM,
+    SupportedBlockchain.BASE,
+    SupportedBlockchain.SCROLL,
+]
+
+SupportedL2WithL1FeesChain = Literal[
+    SupportedBlockchain.OPTIMISM,
+    SupportedBlockchain.BASE,
+    SupportedBlockchain.SCROLL,
+]
+
+L2_CHAIN_IDS_WITH_L1_FEES = [ChainID.OPTIMISM, ChainID.BASE, ChainID.SCROLL]
+
+SupportedL2WithL1FeesChainId = Literal[ChainID.OPTIMISM, ChainID.BASE, ChainID.SCROLL]
```
This can be simplified. No need to repeat yourself.

```suggestion
L2_CHAINS_WITH_L1_FEES_TYPE = Literal[
    SupportedBlockchain.OPTIMISM,
    SupportedBlockchain.BASE,
    SupportedBlockchain.SCROLL,
]
L2_CHAINS_WITH_L1_FEES: tuple[L2_CHAINS_WITH_L1_FEES_TYPE, ...] = typing.get_args(L2_CHAINS_WITH_L1_FEES_TYPE )  # noqa: E501

L2_CHAINIDS_WITH_L1_FEES_TYPE = Literal[ChainID.OPTIMISM, ChainID.BASE, ChainID.SCROLL]
L2_CHAINIDS_WITH_L1_FEES: tuple[L2_CHAINIDS_WITH_L1_FEES_TYPE, ...] = typing.get_args(L2_CHAINIDS_WITH_L1_FEES_TYPE )  # noqa: E501
```

zinkkrysty replied:
oh, nice! I spend half an hour trying to get ChatGPT to suggest me a solution for removing repetition. gave up in the end 😆

LefterisJP commented on rotkehlchen/chain/optimism/types.py:
```diff
@@ -1,53 +1,6 @@
 from typing import TYPE_CHECKING, Literal
 
-from rotkehlchen.types import ChainID, ChecksumEvmAddress, EvmTransaction, EVMTxHash, Timestamp
```
Remove this entire file. It's empty now, isn't it?

zinkkrysty replied:
it's not empty it has:

```
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    # This is a list of all optimism abis that are guaranteed to be returned
    # by EthereumContracts as they are stored in the file/DB
    OPTIMISM_KNOWN_ABI = Literal['VELO_V2_LP']
```

LefterisJP commented on rotkehlchen/chain/scroll/constants.py:
```diff
@@ -18,8 +19,17 @@
 
 PRUNED_NODE_CHECK_TX_HASH: Final = deserialize_evm_tx_hash('0x93a725e6281bd39c6d3798bb818f280b65e5b40bc017a16ed3d8ad5a29fa636c')  # noqa: E501
+
+CPT_SCROLL = 'scroll'
+
+SCROLL_CPT_DETAILS = CounterpartyDetails(
```
Add Final

```suggestion
CPT_SCROLL: Final = 'scroll'
SCROLL_CPT_DETAILS: Final = CounterpartyDetails(
```

LefterisJP commented on rotkehlchen/chain/scroll/constants.py:
```diff
@@ -18,8 +19,17 @@
     active=True,
 )
 
+
```
Do not add unnecessary newlines

```suggestion
```

LefterisJP commented on rotkehlchen/chain/scroll/decoding/decoder.py:
```diff
@@ -42,6 +43,7 @@ def __init__(
                 is_non_conformant_erc721_fn=self._is_non_conformant_erc721,
                 address_is_exchange_fn=self._address_is_exchange,
             ),
+            dbevmtx_class=DBOptimismTx,
```
Is this needed? It's already the default argument no?

Also shouldnt' it be rename to L1WithL2FeesTx?

LefterisJP commented on rotkehlchen/chain/scroll/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,224 @@
+L2_USDC_GATEWAY: Final = string_to_evm_address('0x33B60d5Dd260d453cAC3782b0bDC01ce84672142')
+L2_MESSENGER_PROXY: Final = string_to_evm_address('0x781e90f1c8Fc4611c9b7497C3B47F99Ef6969CbC')
+
+# Topics
+# FinalizeDepositETH - 0x9e86c356e14e24e26e3ce769bf8b87de38e0faa0ed0ca946fa09659aa606bd2d
+FINALIZE_DEPOSIT_ETH: Final = b'\x9e\x86\xc3V\xe1N$\xe2n<\xe7i\xbf\x8b\x87\xde8\xe0\xfa\xa0\xed\x0c\xa9F\xfa\te\x9a\xa6\x06\xbd-'  # noqa: E501
+# WithdrawETH - 0xd8ed6eaa9a7a8980d7901e911fde6686810b989d3082182d1d3a3df6306ce20e
+WITHDRAW_ETH: Final = b'\xd8\xedn\xaa\x9az\x89\x80\xd7\x90\x1e\x91\x1f\xdef\x86\x81\x0b\x98\x9d0\x82\x18-\x1d:=\xf60l\xe2\x0e'  # noqa: E501
+# FinalizeDepositERC20 - 0x165ba69f6ab40c50cade6f65431801e5f9c7d7830b7545391920db039133ba34
+FINALIZE_DEPOSIT_ERC20: Final = b'\x16[\xa6\x9fj\xb4\x0cP\xca\xdeoeC\x18\x01\xe5\xf9\xc7\xd7\x83\x0buE9\x19 \xdb\x03\x913\xba4'  # noqa: E501
+# WithdrawERC20 - 0xd8d3a3f4ab95694bef40475997598bcf8acd3ed9617a4c1013795429414c27e8
+WITHDRAW_ERC20: Final = b"\xd8\xd3\xa3\xf4\xab\x95iK\xef@GY\x97Y\x8b\xcf\x8a\xcd>\xd9azL\x10\x13yT)AL'\xe8"  # noqa: E501
+# RelayedMessage - 0x4641df4a962071e12719d8c8c8e5ac7fc4d97b927346a3d7a335b1f7517e133c
+RELAYED_MESSAGE: Final = b"FA\xdfJ\x96 q\xe1'\x19\xd8\xc8\xc8\xe5\xac\x7f\xc4\xd9{\x92sF\xa3\xd7\xa35\xb1\xf7Q~\x13<"  # noqa: E501
```
Same as in the other file. Remove unnecessary comments
```suggestion
# Topics
FINALIZE_DEPOSIT_ETH: Final = b'\x9e\x86\xc3V\xe1N$\xe2n<\xe7i\xbf\x8b\x87\xde8\xe0\xfa\xa0\xed\x0c\xa9F\xfa\te\x9a\xa6\x06\xbd-'  # noqa: E501
WITHDRAW_ETH: Final = b'\xd8\xedn\xaa\x9az\x89\x80\xd7\x90\x1e\x91\x1f\xdef\x86\x81\x0b\x98\x9d0\x82\x18-\x1d:=\xf60l\xe2\x0e'  # noqa: E501
FINALIZE_DEPOSIT_ERC20: Final = b'\x16[\xa6\x9fj\xb4\x0cP\xca\xdeoeC\x18\x01\xe5\xf9\xc7\xd7\x83\x0buE9\x19 \xdb\x03\x913\xba4'  # noqa: E501
WITHDRAW_ERC20: Final = b"\xd8\xd3\xa3\xf4\xab\x95iK\xef@GY\x97Y\x8b\xcf\x8a\xcd>\xd9azL\x10\x13yT)AL'\xe8"  # noqa: E501
RELAYED_MESSAGE: Final = b"FA\xdfJ\x96 q\xe1'\x19\xd8\xc8\xc8\xe5\xac\x7f\xc4\xd9{\x92sF\xa3\xd7\xa35\xb1\xf7Q~\x13<"  # noqa: E501
```

LefterisJP commented on rotkehlchen/chain/scroll/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,224 @@
+            if (
+                event.event_type == expected_event_type and
+                event.location_label == to_address and
+                # Address for USDC is the gateway's address address
```
```suggestion
                # Address for USDC is the gateway's address
```

LefterisJP commented on rotkehlchen/chain/scroll/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,224 @@
+            ['address', 'address', 'uint256'],
+            method_input_data,
+        )
+        from_address = string_to_evm_address(to_checksum_address(from_hex))
```
you don't need to do string_to_evm_address. Just call our deserialize function directly. `deserialize_evm_address`. That does a to_checksum, underneath.

For this and all other ones

```suggestion
        from_address = deserialize_evm_address(from_hex)
```

LefterisJP commented on rotkehlchen/db/optimismtx.py:
```diff
@@ -21,10 +21,13 @@ class DBOptimismTx(DBEvmTx):
             self,
             write_cursor: 'DBCursor',
-            evm_transactions: list[OptimismTransaction],  # type: ignore[override]
+            evm_transactions: list[L2WithL1FeesTransaction],  # type: ignore[override]
```
Either in this or another PR this class can be renamed as also mentioned above before.

LefterisJP commented on rotkehlchen/chain/ethereum/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,333 @@
+            asset=A_ETH,
+            balance=Balance(amount=fee_amount),
+            location_label=sender,
+            notes=f'Spend {fee_amount} ETH on bridging to Scroll (fees on L2)',
```
```suggestion
            notes=f'Spend {fee_amount} ETH as a fee for bridging to Scrol',
```

LefterisJP commented on rotkehlchen/tests/unit/decoders/test_scroll_bridge.py:
```diff
@@ -0,0 +1,474 @@
+        database=database,
+        tx_hash=evmhash,
+    )
+    user_address = ethereum_accounts[0]
```
For all the tests in this file it would be nice and would save space if you took common attributes out as variables. Such as timestamp, gas, amount etc. And reuse it in the events below.

Can see it in many other decoder tests such as this one: https://github.com/rotki/rotki/blob/341b2ed4c1926190b40f5c748d7074192eab76c9/rotkehlchen/tests/unit/decoders/test_gnosis_sdai.py#L32

zinkkrysty commented on rotkehlchen/chain/ethereum/modules/scroll_bridge/decoder.py:
```diff
@@ -0,0 +1,333 @@
+                event.event_type == HistoryEventType.SPEND and
+                event.asset == A_ETH and
+                event.address == L1_GATEWAY_ROUTER
+            ):
+                # Update the fee event by subtracting the refund amount
```
> Shouldn't this also check if this needs to be done. Aka refund_amount != ZERO?

No, in that case it is subtracting 0 from the fee, which is fine. If no refund then the fee amount is the amount in the fee event. I haven't seen such a case in reality, but I think the code covers it.

LefterisJP reviewed (approved):
LGTM. For the failing test it's just a remote error on a non-vcred test
